gson/src/main/java/com/google/gson/JsonSchemaValidator.java

```diff
@@ -0,0 +1,259 @@
+        }
+
+    }
+
+    static void validateObject(JsonObject schemaNode) throws SchemaValidationException {
+        try {
+            JsonObject childNode = schemaNode.getAsJsonObject("properties");
+            HashMap<String, String> propertyFields = new HashMap<>();
+            for (String key : childNode.keySet()) {
+                if(validateSchemaRootType(childNode.get(key)))
+                    continue;
+                validateSchemaNode(childNode.get(key));
+                propertyFields.put(key, key);
+            }
+            if(schemaNode.has("required")) {
+                JsonArray requiredFields = schemaNode.getAsJsonArray("required");
+                for (JsonElement element : requiredFields) {
+                    assert(propertyFields.containsKey(element.getAsString()));
+                }
+            }
+        } catch (SchemaValidationException ex) {
+            throw ex;
+        } catch (AssertionError ex) {
+            throw new SchemaValidationException("Required properties not present in schema");
+        } catch (Throwable ex) {
+            throw new SchemaValidationException(ex);
+        }
+    }
+
+    static void validateArray(JsonObject schemaNode) throws SchemaValidationException {
+        if (schemaNode.has("items")){
+            validateSchemaNode(schemaNode.getAsJsonObject("items"));
+        }
+        if (schemaNode.has("contains")){
+            validateSchemaNode(schemaNode.getAsJsonObject("contains"));
+        }
+        if ((!schemaNode.has("items") || schemaNode.getAsJsonObject("items").keySet().isEmpty()) &&
+                !schemaNode.has("contains")){
+            throw new SchemaValidationException("Empty \"items\" object requires \"contains\" keyword");
+        }
+
+        if (schemaNode.has("prefixItems")){
+            JsonArray prefixItemsArray = schemaNode.getAsJsonArray("prefixItems");
+            if (prefixItemsArray.isEmpty()){
+                throw new SchemaValidationException("\"prefixItems\" may not be empty");
+            }
+            for (JsonElement el : prefixItemsArray){
+                validateSchemaNode(el.getAsJsonObject());
+            }
+        }
+
+        verifyAndGetFieldOrDefault(schemaNode, "minItems", JsonElement::getAsInt,  0);
+        verifyAndGetFieldOrDefault(schemaNode, "uniqueItems", JsonElement::getAsBoolean, true);
+
+        if (schemaNode.has("uniqueItems")){
+            try {
+                schemaNode.getAsJsonPrimitive("uniqueItems").getAsBoolean();
+            } catch (Throwable throwable){
+                throw new SchemaValidationException("\"uniqueItems\" must be a boolean", throwable);
+            }
+        }
+
+    }
+
+    static void validateTypeField(JsonObject schema) throws SchemaValidationException {
+        JsonElement type = schema.get("type");
+        if (type == null){
+            throw new SchemaValidationException("Type not defined");
+        }
```

Comment:
I think the schema object is not required to contain the `type` property.

Reply:
@alevarn This is related to issue #6 I think, because it can either contain the `type` property or a `$ref` property to an external schema.